src/main/java/jenkins/scm/api/SCMHeadMixinEqualityGenerator.java

```diff
@@ -0,0 +1,488 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package jenkins.scm.api;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.WeakHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import jenkins.scm.api.mixin.SCMHeadMixin;
+import net.jcip.annotations.GuardedBy;
+import org.apache.commons.lang.ClassUtils;
+import org.kohsuke.asm5.ClassWriter;
+import org.kohsuke.asm5.Label;
+import org.kohsuke.asm5.MethodVisitor;
+import org.kohsuke.asm5.Opcodes;
+import org.kohsuke.asm5.Type;
+
+import static org.kohsuke.asm5.Opcodes.ACC_PUBLIC;
+import static org.kohsuke.asm5.Opcodes.ALOAD;
+import static org.kohsuke.asm5.Opcodes.ASTORE;
+import static org.kohsuke.asm5.Opcodes.CHECKCAST;
+import static org.kohsuke.asm5.Opcodes.DCMPL;
+import static org.kohsuke.asm5.Opcodes.DLOAD;
+import static org.kohsuke.asm5.Opcodes.DSTORE;
+import static org.kohsuke.asm5.Opcodes.FCMPL;
+import static org.kohsuke.asm5.Opcodes.FLOAD;
+import static org.kohsuke.asm5.Opcodes.FSTORE;
+import static org.kohsuke.asm5.Opcodes.GOTO;
+import static org.kohsuke.asm5.Opcodes.ICONST_0;
+import static org.kohsuke.asm5.Opcodes.ICONST_1;
+import static org.kohsuke.asm5.Opcodes.IFEQ;
+import static org.kohsuke.asm5.Opcodes.IFNE;
+import static org.kohsuke.asm5.Opcodes.IFNONNULL;
+import static org.kohsuke.asm5.Opcodes.IFNULL;
+import static org.kohsuke.asm5.Opcodes.IF_ICMPEQ;
+import static org.kohsuke.asm5.Opcodes.ILOAD;
+import static org.kohsuke.asm5.Opcodes.INVOKEINTERFACE;
+import static org.kohsuke.asm5.Opcodes.INVOKESPECIAL;
+import static org.kohsuke.asm5.Opcodes.INVOKEVIRTUAL;
+import static org.kohsuke.asm5.Opcodes.IRETURN;
+import static org.kohsuke.asm5.Opcodes.ISTORE;
+import static org.kohsuke.asm5.Opcodes.LCMP;
+import static org.kohsuke.asm5.Opcodes.LLOAD;
+import static org.kohsuke.asm5.Opcodes.LSTORE;
+import static org.kohsuke.asm5.Opcodes.RETURN;
+
+/**
+ * Generates {@link SCMHeadMixin.Equality} instances for concrete {@link SCMHead} instance types.
+ * We need {@link SCMHead} instances to perform equality based on the {@link SCMHead#getName()} plus all the property
+ * values declared on the {@link SCMHeadMixin} interfaces implemented by the {@link SCMHead} concrete type.
+ * As {@link SCMHead#equals(Object)} is expected to be a hot method, we'd much rather avoid using reflection, so
+ * instead we use bytecode generation to create our {@link SCMHeadMixin.Equality} subclass for us.
```

Comment:
Well, I'm not sure but I think I'd prefer fixing any eventual problems in those plugins than trying to figure out a future bug in this code :)